packages/admin/src/common/toolbar/Toolbar.styles.tsx

```diff
@@ -0,0 +1,62 @@
+import { Theme } from "@material-ui/core/styles";
+import { makeStyles } from "@material-ui/styles";
+
+import { useComponentThemeProps } from "../../mui/useComponentThemeProps";
+
+export interface ToolbarThemeProps {
+    elevation?: number;
+}
+
+export type CometAdminToolbarClassKeys = "root" | "muiToolbar" | "mainContentContainer" | "actionContainer";
+
+export const useStyles = makeStyles<Theme, { headerHeight: number }, CometAdminToolbarClassKeys>(
+    () => ({
+        root: {
+            position: "sticky",
+            display: "flex",
+            flexDirection: "column",
+            justifyContent: "center",
+            top: (props) => props.headerHeight,
+            padding: 0,
+            "& [class*='MuiButton']": {
+                textTransform: "none",
+            },
+        },
+        muiToolbar: {
+            display: "flex",
+            flex: 1,
+            alignItems: "stretch",
+        },
+        mainContentContainer: {
+            display: "flex",
+            flex: 1,
+        },
+        actionContainer: {
+            display: "flex",
+            alignItems: "center",
+            /* Margin on all first level children, without last one */
+            "& > *:not(:last-child)": {
+                marginRight: 20,
+            },
+        },
+    }),
+    { name: "CometAdminToolbar" },
+);
+
+export function useThemeProps() {
+    const { elevation = 4, ...restProps } = useComponentThemeProps<ToolbarThemeProps>("CometAdminToolbar") ?? {};
+    return { elevation, ...restProps };
+}
+
+// Theme Augmentation
+declare module "@material-ui/core/styles/overrides" {
+    interface ComponentNameToClassKey {
+        CometAdminToolbar: CometAdminToolbarClassKeys;
+    }
+}
+
+declare module "@material-ui/core/styles/props" {
+    interface ComponentsPropsList {
+        CometAdminToolbar: ToolbarThemeProps;
+    }
+}
```

Comment:
👍🏻

Reply:
@rabengraph @nsams maybe a short statement to this kind of theme approach we choose.

i think @jamesricky did a great job in setting up the comet theme and the initial structuring of the theme repo. Further more, we (@jamesricky and @manuelblum) talked a bit about Theming, structuring the things and enhancing some portions of the whole theming process. I want to show you a short overview what we are trying to achieve and what our outcome was:

### Keep the Augmentation on Component Level (same file)

Currently we have all theme augmentation in the [root of @comet/admin package](https://github.com/vivid-planet/comet-admin/blob/next/packages/admin/src/themeAugmentation.ts). This can be a really pain to make custom components theme able with overrides and pass custom theme props to it because there is no real validation on the key level. It is also a common point of conflicts. So we tried to put the Theme Augmentation to the component itself. At first i tried to put all jsx rendering, logic and styling into one file and it got a bit confusing and big after some time. So we split up the Component and styling to two separat files (fur sure this can also be done in one file ;-) but i think this kind offers a bit more structuring). 



### Interplay between Theme Augmentation, Props, and Overrides

**Theme Augmentation**
- themeAugmentation: Is a Technik to add some declaration to existing Modules/Components.
     - In our case we currently focus on customizing declarations for `ComponentNameToClassKey` and `ComponentsPropsList` from Material UI so we can override this kind of things in our custom @comet/admin-theme
     - More infos about theme augmentation can be found here: [Material UI - Customization of theme](https://material-ui.com/guides/typescript/#customization-of-theme) and [Typescript  Module Augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation)

**Overrides**
The Overridable classes have to be declared in the Theme Augmentation [Toolbar.styles.tsx](https://github.com/vivid-planet/comet-admin/blob/toolbar/packages/admin/src/common/toolbar/Toolbar.styles.tsx#L45)
This classes can then be overridden in any theme, e.g. [@comet/admin-theme](https://github.com/vivid-planet/comet-admin/blob/toolbar/packages/admin-theme/src/cometAdminOverrides/toolbar.ts)

**Props**
Sometimes it is also neccessary to set some custom props per theme (we will call them `ComponentThemeProps`). All available `ComponentThemeProps` must be also [declared in the Theme Augmentation](https://github.com/vivid-planet/comet-admin/blob/toolbar/packages/admin/src/common/toolbar/Toolbar.styles.tsx#L52).

To access this `ComponentThemeProps` we built a small hook `useComponentThemeProps` which ensures that only valid strings (based on all available themeAugmentations) can be passed, and return the theme customized componentprops. This props can then easily be specified in a theme e.g.: [Toolbar](https://github.com/vivid-planet/comet-admin/blob/toolbar/packages/admin-theme/src/cometAdminProps/toolbar.tsx) or [Toolbar Back - Icon](https://github.com/vivid-planet/comet-admin/blob/toolbar/packages/admin-theme/src/cometAdminProps/toolbarBackButton.tsx)


@jamesricky have missed something or do you want to add something?

Reply:
@manuelblum i think you covered it all 👍🏻

Reply:
@manuelblum good overview! We should put this somewhere where we can find it again - docs?